pkg/controller/networkpolicy/clusternetworkpolicy.go

```diff
@@ -51,5 +109,159 @@ func (n *NetworkPolicyController) deleteCNP(old interface{}) {
+			if err != nil {
+				klog.Errorf("Failure processing ClusterNetworkPolicy %s IPBlock %v: %v", cnp.Name, peer.IPBlock, err)
+				continue
+			}
+			ipBlocks = append(ipBlocks, *ipBlock)
+		} else {
+			normalizedUID := n.createAddressGroupForCRD(peer, cnp)
+			addressGroups = append(addressGroups, normalizedUID)
+		}
+	}
+	return &networking.NetworkPolicyPeer{AddressGroups: addressGroups, IPBlocks: ipBlocks}
+}
+
+// createAddressGroupForCRD creates an AddressGroup object corresponding to a
+// secv1alpha1.NetworkPolicyPeer object in Cluster NetworkPolicyRule. This
+// function simply creates the object without actually populating the
+// PodAddresses as the affected Pods are calculated during sync process.
+func (n *NetworkPolicyController) createAddressGroupForCRD(peer secv1alpha1.NetworkPolicyPeer, np *secv1alpha1.ClusterNetworkPolicy) string {
+	groupSelector := toGroupSelector("", peer.PodSelector, peer.NamespaceSelector)
+	normalizedUID := getNormalizedUID(groupSelector.NormalizedName)
+	// Get or create an AddressGroup for the generated UID.
+	_, found, _ := n.addressGroupStore.Get(normalizedUID)
+	if found {
+		return normalizedUID
+	}
+	// Create an AddressGroup object per Peer object.
+	addressGroup := &antreatypes.AddressGroup{
+		UID:      types.UID(normalizedUID),
+		Name:     normalizedUID,
+		Selector: *groupSelector,
+	}
+	klog.V(2).Infof("Creating new AddressGroup %s with selector (%s)", addressGroup.Name, addressGroup.Selector.NormalizedName)
+	n.addressGroupStore.Create(addressGroup)
+	return normalizedUID
+}
+
+// processClusterNetworkPolicy creates an internal NetworkPolicy instance
+// corresponding to the secv1alpha1.ClusterNetworkPolicy object. This method
+// does not commit the internal NetworkPolicy in store, instead returns an
+// instance to the caller wherein, it will be either stored as a new Object
+// in case of ADD event or modified and store the updated instance, in case
+// of an UPDATE event.
+func (n *NetworkPolicyController) processClusterNetworkPolicy(cnp *secv1alpha1.ClusterNetworkPolicy) *antreatypes.NetworkPolicy {
+	appliedToGroupNames := make([]string, 0, len(cnp.Spec.AppliedTo))
+	// Create AppliedToGroup for each AppliedTo present in
+	// ClusterNetworkPolicy spec.
+	for _, at := range cnp.Spec.AppliedTo {
+		appliedToGroupNames = append(appliedToGroupNames, n.createAppliedToGroup("", at.PodSelector, at.NamespaceSelector))
+	}
+	rules := make([]networking.NetworkPolicyRule, 0, len(cnp.Spec.Ingress)+len(cnp.Spec.Egress))
+	// Compute NetworkPolicyRule for Egress Rule.
+	for idx, ingressRule := range cnp.Spec.Ingress {
+		// Set default action to ALLOW to allow traffic.
+		action := ingressRule.Action.ToUpper()
+		rules = append(rules, networking.NetworkPolicyRule{
+			Direction: networking.DirectionIn,
+			From:      *n.toAntreaPeerForCRD(ingressRule.From, cnp, networking.DirectionIn),
+			Services:  toAntreaServicesForCRD(ingressRule.Ports),
+			Action:    &action,
+			Priority:  int32(idx),
```

Comment:
It stores the index of rule within ingress/egress rules.

Reply:
But why we need to set it in controller? Agent also can know the indexes from the rule slice?

Reply:
Decision to set the index in controller is made by @abhiraut, theoretically I agree it can be inferred from the agent, if we ensure that the internal NP is created with rules with the same order and those don't get shuffled in the networking channel.